### ISSUE TITLE
should use a port other than 80 in test_ipv6

### ISSUE THREAD

**Issue by brianjmurrell** [sabnzbd/sabnzbd#274]
I have discovered that ```test_ipv6()``` will fail even if IPv6 is available if SABzbd is going through a transparent proxy.  Since port 443 is not proxyable, it would be a better choice for the test than port 80.

**Comment by shypike** [comment]
Port 443 isn't proxyable? So your proxy doesn't support https?

**Comment by brianjmurrell** [comment]
Port 443 is not transparently proxyable, no and that makes it an excellent candidate for ```test_ipv6()```.

**Comment by sanderjo** [comment]
Question about your transparant proxy: does it proxy NNTP / News traffic?

**Comment by brianjmurrell** [comment]
No.

**Comment by sanderjo** [comment]
If the proxy does not proxy NNTP / News traffic, can SABnzbd download stuff from NNTP servers? If not, what is the use of SABnzbd behind that proxy?

**Comment by brianjmurrell** [comment]
Yes, SABnzbd can download stuff from NNTP servers.  But TBH, I'm not sure where you are going with this line of inquiry.  It seems irrelevant to the problem at hand of using port 80 for the "do we have ipv6" test since 80 can be intercepted for transparent proxying.

**Comment by sanderjo** [comment]
Because I don't understand your setup nor use case.

**Comment by brianjmurrell** [comment]
It's quite simple.  All hosts go through a gateway to the Internet.  That gateway is a transparent proxy for port 80.  Maybe the problem is that you don't quite understand how a transparent proxy works.  Here's an [explanation](https://en.wikipedia.org/wiki/Proxy_server#Transparent_proxy).

**Comment by shypike** [comment]
My takeaway from the Wikipedia explanation is this: "Clients need not be aware of the existence of the proxy".
So it seems that the proxy in case is semi-transparent. It is ONLY transparent for port 80, but not for any other port (not even 443?). Is this a common situation?

**Comment by brianjmurrell** [comment]
It's completely pass-through for any/every other port other than 80 since it cannot be even semi-transparent for ports like 443 because HTTPS is unproxyable (well, without distributing CA certificates and launching MitM attacks on your users -- which I am sure some corporations do but I am not interested in that here).

**Comment by shypike** [comment]
@brianjmurrell 
OK, so it intercepts only HTTP (port 80) requests, but everything else gets through?
That would be an easy one to ignore: just use HTTPS for everything?
I'm just asking, because I never heard of such a proxy.
The (corporate) proxies I know just block everything, except HTTP and HTTPS requests and they still don't need a MTM attack. Of course these aren't transparent but do require you to setup a proxy in the web browser.
So using a port 443 check instead of a port 80 would solve the issue? Can we assume that all hosts respond the same to 443 as they would do to 80?
@sanderjo : can you confirm this?

**Comment by sanderjo** [comment]
I still don't understand: If there is a transparant proxy on port 80, why is a connect to a port 80 site not working? Is it a problem because of IPv6? Is it a problem in the transparant proxy? If a RSS-index site is IPv6 enabled, is there a probleem too (which would make solving the above problem silly)? 
Questions, questions, questions

AFAIK, connecting via port 443 requires ssl and valid certifcates, That could be a problem for other systems; low end, or incorrect certificate setup. So it's a bit like "being bitten by the cat, or the dog" ;-)

**Comment by brianjmurrell** [comment]
@sanderjo No.  There is no SSL needed to test IPv6 connectivity using port 443 rather than port 80.  The test does not actually do anything beyond verify that a socket can be opened.  No payload is actually sent/received so there is no need to do anything with SSL.  Simply s/80/443/ and the test will connect to 443 and if it connects, then the test is successful.

**Comment by brianjmurrell** [comment]
@shypike Yes, it only intercepts port 80.  The point of such a proxy is not necessarily to wall people in but to allow local caching for efficiency and to allow site-wide ad/tracker blocking.

And yes, as more and more traffic becomes HTTPS this will become less and less useful but there is still enough HTTP to make it useful.

As for all hosts using 443 in addition to 80, we don't need to worry about all of them, just the one(s) being used in ```test_ipv6() ``` which are the google server(s) and we know they use HTTPS in addition to HTTP.

**Comment by brianjmurrell** [comment]
Why was this closed?  The issue was not resolved was it?

**Comment by shypike** [comment]
We were a bit too much in the cleaning mood.

**Comment by Safihre** [comment]
But are we ever going to implement this? I marked this for closure because of the discussion above and that only this very specific use-case would need a different port than 80. What can be more generally applicable than port 80?

**Comment by shypike** [comment]
@Safihre Reading back Sander's comment, it doesn't seem to have side effects when using 443 instead of 80.

**Comment by Safihre** [comment]
@sanderjo so what do you think? Just change it to 443? 
I tend to wonder the same as you:
>  Are we solving a problem in SAB, or working around a problem in a certain environment?

**Comment by brianjmurrell** [comment]
The thing is, that the environment is not entirely unique, it's just leading edge.  Transparent proxies exist in many places.  Most people don't see or recognize them.  It's just that this place is on the leading edge, using transparent proxies and making IPv6 available.  It's the general lack of IPv6 deployment that is making this issue look so unique.

As more and more IPv6 becomes available, this problem will surface more frequently.

**Comment by sanderjo** [comment]
@brianjmurrell 
On what platform do you run SABnzbd: Linux (thus from source), OSX or Windows?

**Comment by brianjmurrell** [comment]
@sanderjo: Linux.

**Comment by sanderjo** [comment]
@brianjmurrell Good. That means you can test stuff from source code.

I'll be back.

**Comment by sanderjo** [comment]
I had a look at the test_ipv6() code, and it does not use a port. It's a socket connect, and at that low level there are no ports involved.

Yes, there is this code:
```
info = socket.getaddrinfo(cfg.ipv6_test_host(), 80, socket.AF_INET6, socket.SOCK_STREAM, socket.IPPROTO_IP, socket.AI_CANONNAME)
```
with '80', but that is just the DNS lookup. You can change that "80" to anything: 443, 0, 11111

So ... I guess there is no solution for you. To verify: in sabnzbd/__init__.py can you find that line and change the 80 to 0 or 443 or 11111 and verify you get NO other result?

However I wrote some code for testing on your system.

Result with correct working IPv6:
```
$ python test_ipv6.py 
ipv6testhost is www.google.com
canonname ('2a00:1450:400c:c09::68', 443, 0, 0)
socket connect: IPv6 test successful. Enabling IPv6
Yes, IPv6 connection via HTTP http://[2a00:1450:400c:c09::68]/
Yes, IPv6 connection via HTTPS https://[2a00:1450:400c:c09::68]/
Done
```
All results are the same and consistent.

No working IPv6:
```
$ python test_ipv6.py 
ipv6testhost is www.google.com
canonname ('2a00:1450:400c:c09::68', 443, 0, 0)
socket connect: Cannot reach IPv6 test host. Disabling IPv6
No IPv6 connection via HTTP http://[2a00:1450:400c:c09::68]/
No IPv6 connection via HTTPS https://[2a00:1450:400c:c09::68]/
Done
```
Again all results are the same and consistent.

Code is here: https://raw.githubusercontent.com/sanderjo/test_ipv6/master/test_ipv6.py

Run it on your system behind the (non)transparant proxy and post the result.

**Comment by brianjmurrell** [comment]
I have 0.7.20 here.  In ```sabnzbd/newswrapper.py``` I have:

```
def test_ipv6():
    """ Check if external IPv6 addresses are reachable """
    # Use google.com to test IPv6 access
    try:
        info = socket.getaddrinfo('www.google.com', 443, socket.AF_INET6, socket.SOCK_STREAM,
                                  socket.IPPROTO_IP, socket.AI_CANONNAME)
    except:
        return False

    try:
        af, socktype, proto, canonname, sa = info[0]
        sock = socket.socket(af, socktype, proto)
        sock.settimeout(6)
        sock.connect(sa[0:2])
        sock.close()
        return True
    except:
        return False

_EXTERNAL_IPV6 = test_ipv6()
```

That sure looks like it's making a socket connection to me.  And notice I changed it from 80 to 443 to make it work.  When that 443 is 80 it doesn't work.

Oh wait.  I see you edited your comment since I got my e-mail copy to remove the part about it not connecting to a socket.  So you agree that it is making a socket connection to port 80, yes?

Your test script's output on my system here:

```
$ python test_ipv6.py 
ipv6testhost is ipv6.google.com
Loop:
canonname ('2607:f8b0:400b:80a::1001', 80, 0, 0)
socket connect: Cannot reach IPv6 test host.
No IPv6 connection via HTTP http://[2607:f8b0:400b:80a::1001]/
No IPv6 connection via HTTPS https://[2607:f8b0:400b:80a::1001]/
canonname ('2607:f8b0:400b:80a::1001', 443, 0, 0)
socket connect: IPv6 test successful.
No IPv6 connection via HTTP http://[2607:f8b0:400b:80a::1001]/
No IPv6 connection via HTTPS https://[2607:f8b0:400b:80a::1001]/
Done
```

**Comment by Safihre** [comment]
So, 443 it is then it seems! Or can we think of some other exotic setup where that would fail? @sander

**Comment by sanderjo** [comment]
With the script I wrote, change the first line to
```
ipv6testhost = "test-ipv6.sabnzbd.org"
```
(as that is test host that is used in SAB 0.8.0) and run it again. I did that, and I don't understand the result.

**Comment by sanderjo** [comment]
Furthermore:
```
No IPv6 connection via HTTP http://[2607:f8b0:400b:80a::1001]/
No IPv6 connection via HTTPS https://[2607:f8b0:400b:80a::1001]/
```
That seems you have no IPv6 connectivity at all. Do you agree?

Can your SABnzbd connect to newszilla6.xs4all.nl (no account needed).

**Comment by brianjmurrell** [comment]
```
ipv6testhost is test-ipv6.sabnzbd.org
Loop:
canonname ('2400:cb00:2048:1::6812:28d7', 80, 0, 0)
socket connect: Cannot reach IPv6 test host.
No IPv6 connection via HTTP http://[2400:cb00:2048:1::6812:28d7]/
No IPv6 connection via HTTPS https://[2400:cb00:2048:1::6812:28d7]/
Loop:
canonname ('2400:cb00:2048:1::6812:29d7', 80, 0, 0)
socket connect: Cannot reach IPv6 test host.
No IPv6 connection via HTTP http://[2400:cb00:2048:1::6812:29d7]/
No IPv6 connection via HTTPS https://[2400:cb00:2048:1::6812:29d7]/
canonname ('2400:cb00:2048:1::6812:28d7', 443, 0, 0)
socket connect: IPv6 test successful.
No IPv6 connection via HTTP http://[2400:cb00:2048:1::6812:28d7]/
No IPv6 connection via HTTPS https://[2400:cb00:2048:1::6812:28d7]/
canonname ('2400:cb00:2048:1::6812:29d7', 443, 0, 0)
socket connect: IPv6 test successful.
No IPv6 connection via HTTP http://[2400:cb00:2048:1::6812:29d7]/
No IPv6 connection via HTTPS https://[2400:cb00:2048:1::6812:29d7]/
Done
```

Here's the ```tcpdump``` output of running the test script:

```
10:04:21.223295 IP6 my:ipv6:host.57456 > 2400:cb00:2048:1::6812:29d7.443: Flags [S], seq 2547362499, win 24400, options [mss 1220,sackOK,TS val 29217843 ecr 0,nop,wscale 7], length 0
10:04:21.232693 IP6 2400:cb00:2048:1::6812:29d7.443 > my:ipv6:host.57456: Flags [S.], seq 3664956040, ack 2547362500, win 24400, options [mss 1220,nop,nop,sackOK,nop,wscale 10], length 0
10:04:21.233003 IP6 my:ipv6:host.57456 > 2400:cb00:2048:1::6812:29d7.443: Flags [.], ack 1, win 191, length 0
10:04:21.233098 IP6 my:ipv6:host.57456 > 2400:cb00:2048:1::6812:29d7.443: Flags [F.], seq 1, ack 1, win 191, length 0
10:04:21.244400 IP6 2400:cb00:2048:1::6812:29d7.443 > my:ipv6:host.57456: Flags [F.], seq 1, ack 2, win 24, length 0
10:04:21.244685 IP6 my:ipv6:host.57456 > 2400:cb00:2048:1::6812:29d7.443: Flags [.], ack 2, win 191, length 0
```

I most definitely do have IPv6:
```
$ ping6 2607:f8b0:400b:80a::1001
PING 2607:f8b0:400b:80a::1001(2607:f8b0:400b:80a::1001) 56 data bytes
64 bytes from 2607:f8b0:400b:80a::1001: icmp_seq=1 ttl=58 time=10.3 ms
64 bytes from 2607:f8b0:400b:80a::1001: icmp_seq=2 ttl=58 time=11.0 ms
64 bytes from 2607:f8b0:400b:80a::1001: icmp_seq=3 ttl=58 time=10.7 ms

$ ping6 test-ipv6.sabnzbd.org
PING test-ipv6.sabnzbd.org(2400:cb00:2048:1::6812:28d7) 56 data bytes
64 bytes from 2400:cb00:2048:1::6812:28d7: icmp_seq=1 ttl=60 time=10.1 ms
64 bytes from 2400:cb00:2048:1::6812:28d7: icmp_seq=2 ttl=60 time=10.2 ms
64 bytes from 2400:cb00:2048:1::6812:28d7: icmp_seq=3 ttl=60 time=10.4 ms

$ telnet newszilla6.xs4all.nl 119
Trying 2001:888:0:18::119...
Connected to newszilla6.xs4all.nl.
Escape character is '^]'.
201 newszilla.xs4all.nl NNRP Service Ready (dreader37) (no posting)
```

**Comment by thezoggy** [comment]
one must assume your running python and ping6 as the same user on the same box with the same rights? what version of python.. python -V

**Comment by brianjmurrell** [comment]
Yes, all the same user.  ```Python 2.7.10```

**Comment by Safihre** [comment]
Does it matter if the ```urlopen``` succeeds? The socket-connection on port 443 succeeds.
So all in all, 443 seems to be just as safe a choice.
Your script reports full succes for everything here on my IPv6 connection, 80 and 443.

**Comment by shypike** [comment]
Solved by 9457883

**Comment by brianjmurrell** [comment]
Excellent!
